implementing replace_picture()

m-entrup opened the pull request:
We iterate over all shapes `find_shapes()` returns.
Raises a `FileNotFoundError` if `filename` does not point to a file.
The picture is created as a new shape.
The last step is to delete the old shape.

m-entrup committed:
Implementing replace_picture():

We iterate over all shapes find_shapes() return.
Raises a FileNotFoundError if `filename`does not point to a file.
The picture is created as a new shape
The last step is to delete the old shape.

m-entrup committed:
Implementing replace_picture():

We iterate over all shapes find_shapes() return.
Raises a FileNotFoundError if `filename`does not point to a file.
The picture is created as a new shape
The last step is to delete the old shape.

m-entrup committed:
Adding playground code for adding an image shape:

The code uses data/example01.pptx to test adding an image shape.
The placeholder with the text #logo gets replaced by the image.

m-entrup committed:
Merge remote-tracking branch 'upstream/master'

timhoffm committed:
Fix type specification

michaelosthege commented on pptx_blueprint/__init__.py:
```diff
@@ -37,7 +37,27 @@ def replace_picture(self, label: str, filename: _Pathlike) -> None:
+                top=old_shape.top,
+                width=old_shape.width,
+                height=old_shape.height
+            )
```
Should this be a `with open() …`?
Or is that not required when only read access is performed?

m-entrup committed:
Merge remote-tracking branch 'upstream/master'

lysnikolaou commented on pptx_blueprint/__init__.py:
```diff
@@ -37,7 +37,27 @@ def replace_picture(self, label: str, filename: _Pathlike) -> None:
+            filename = pathlib.Path(filename)
+        if not filename.is_file():
+            raise FileNotFoundError(f"The file does not exist: {filename}")
+        img_file = open(filename, "rb")
```
```suggestion
        with open(filename, "rb") as img_file:
```

lysnikolaou commented on pptx_blueprint/__init__.py:
```diff
@@ -37,7 +37,27 @@ def replace_picture(self, label: str, filename: _Pathlike) -> None:
+        if not filename.is_file():
+            raise FileNotFoundError(f"The file does not exist: {filename}")
+        img_file = open(filename, "rb")
+        old_shape: BaseShape
```
I would maybe type `shapes_to_replace`, instead of `old_shape`.

m-entrup replied:
I already use `shapes_to_replace` for the iterator that is returned by `_find_shapes()`. In my opinion the singular version `shape_to_replace` is to close to `shapes_to_replace` and therefor `old_shape` makes it easier to distinguish between the iterator and the single shape.

Commit:
First test

m-entrup committed:
Ignoring pipenv and PyCharm related files:

We use requirements.txt and dev_requirements.txt.
Therefore, Pipfile and Pipfile.lock are not needed.
Ignoring all PyCharm related files in .idea/.

m-entrup committed:
Ignoring pipenv and PyCharm related files:

We use requirements.txt and dev_requirements.txt.
Therefore, Pipfile and Pipfile.lock are not needed.
Ignoring all PyCharm related files in .idea/.

m-entrup committed:
Merge remote-tracking branch 'upstream/master'

m-entrup committed:
Merge remote-tracking branch 'origin/master'

m-entrup committed:
Enhancing testing_1.py:

Running replace_by_image() a second time.
verifying the second run does not create new shapes.

m-entrup committed:
Merge remote-tracking branch 'upstream/master'

cmahr commented on .gitignore:
```diff
@@ -90,7 +90,8 @@ target/
 #   install all needed dependencies.
-#Pipfile.lock
+Pipfile
+Pipfile.lock
```
What's the reason for not committing the dependency lock file? I would only *not* do this if there is a problem...

m-entrup replied:
As we use requirements.txt and dev_requirements.txt for listing dependencies a Pipfile would duplicate the requirements.

cmahr commented on pptx_blueprint/__init__.py:
```diff
@@ -37,7 +37,27 @@ def replace_picture(self, label: str, filename: _Pathlike) -> None:
         """
-        pass
+        shapes_to_replace = self._find_shapes(label)
+        if not shapes_to_replace:
```
This should be an Exception. Don't let errors pass silently!

m-entrup replied:
It is a good idea to handle the problem on a higher level. I will replace `return` by
```python
raise ValueError(f"The label '{label}' can't be found in the template.")
```

cmahr commented on pptx_blueprint/__init__.py:
```diff
@@ -37,7 +37,27 @@ def replace_picture(self, label: str, filename: _Pathlike) -> None:
+        img_file = open(filename, "rb")
+        old_shape: BaseShape
+        for old_shape in shapes_to_replace:
+            slide_shapes = old_shape._parent
```
Not a big fan of using the internals of the pptx library. But if there is really no better way...

timhoffm replied:
It seems there is currently no other solution:
https://github.com/scanny/python-pptx/issues/246

Could be mentionend in a code comment.

cmahr replied:
Thx for linking the resource.

cmahr commented on pptx_blueprint/__init__.py:
```diff
@@ -37,7 +37,27 @@ def replace_picture(self, label: str, filename: _Pathlike) -> None:
+            )
+            # Removing shapes is performed at the lxml level.
+            # The `element` attribute contains an instance of `lxml.etree._Element`.
+            slide_shapes.element.remove(old_shape.element)
```
Yikes! We should really not deal with xml ourselves. We should open a RP in pptx to get this feature implemented. For the time being, we should either 1) isolate this in a private method marked deprecated to prevent excessive use, or 2) decorate the pptx presentation object.

m-entrup committed:
Adding a test for image scaling.

m-entrup committed:
Extending `replace_picture()` to scale images:

Large images are always scaled.
The aspect ratio is kept.
The new kwarg `do_not_scale_up`can disable upscaling.
Centering is always active.

m-entrup committed:
Merge remote-tracking branch 'upstream/master'

m-entrup commented:
This PR is messed up a bit by myself. It should be best to use PR #9 instead. All changes to `replace_image()` are included into PR #9.
